### ISSUE TITLE
Crashes when building with Xcode 16 in release build config

### ISSUE THREAD

**Issue by TheMen4ce** [ernesto-elsaesser/WebArchiver#3]
When building my App with **Xcode 16** in the **release build config** using an **iPhone X with iOS 16.7.10** I get this error:

```
Crashed: WebArchiverWorkQueue (QOS: UNSPECIFIED)
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000114f10020
```

This error doesn't occur when building with **Xcode 15.4**, nor does it when using the **debug build config**, nor does it when using an iPhone 16 with iOS 18.0.0.
Additionally, this error doesn't happen every time but only every 2nd app start on average (!!)

I have no idea what causes this issue nor if it will be fixed in a future version of Xcode. It took me quite a while to realize that these crashes were not related to my changes but to the version of Xcode used to build the app. 

I post this here as an info for other people having the same issue. 

If anyone finds a way on how to properly fix this for Xcode 16 – please let me know 🙏

```
Crashed: WebArchiverWorkQueue (QOS: UNSPECIFIED)
0  libsystem_platform.dylib       0xa84 _platform_memmove + 52
1  libxml2.2.dylib                0xf70 xmlBufAdd + 220
2  libxml2.2.dylib                0x11930 xmlParserInputBufferCreateMem + 96
3  libxml2.2.dylib                0x14b4 xmlCreateMemoryParserCtxt + 72
4  libxml2.2.dylib                0x2e74c htmlReadMemory + 60
5  myApp                        0xa4924 XMLDocument.__allocating_init(string:encoding:) + 130 (Document.swift:130)
6  myApp                        0xa3738 specialized static WebArchiver.extractHTMLReferences(from:includeJavascript:) + 86 (WebArchiver.swift:86)
7  myApp                        0xa0a60 closure #1 in static WebArchiver.archive(url:cookies:includeJavascript:skipCache:completion:) + 23 (WebArchive.swift:23)
8  myApp                        0xa0e30 partial apply for closure #1 in static WebArchiver.archive(url:cookies:includeJavascript:skipCache:completion:) + 4374859312 (<compiler-generated>:4374859312)
9  myApp                        0x9c844 closure #1 in ArchivingSession.load(url:fallback:expand:) + 51 (ArchivingSession.swift:51)
10 myApp                        0x9d480 partial apply for closure #1 in ArchivingSession.load(url:fallback:expand:) + 4374844544 (<compiler-generated>:4374844544)
11 myApp                        0x993f0 thunk for @escaping @callee_guaranteed @Sendable (@guaranteed Data?, @guaranteed NSURLResponse?, @guaranteed Error?) -> () + 4374828016 (<compiler-generated>:4374828016)
12 CFNetwork                      0x22b6c CFURLRequestSetMainDocumentURL + 3000
13 CFNetwork                      0x33d80 _CFNetworkErrorCopyLocalizedDescriptionWithHostname + 11944
14 Foundation                     0x51bfc __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 24
15 Foundation                     0x63f74 -[NSBlockOperation main] + 104
16 Foundation                     0x3c1f4 __NSOPERATION_IS_INVOKING_MAIN__ + 24
17 Foundation                     0x4d5bc -[NSOperation start] + 788
18 Foundation                     0x50c48 __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 24
19 Foundation                     0x5edf0 __NSOQSchedule_f + 184
20 libdispatch.dylib              0x12830 _dispatch_block_async_invoke2 + 148
21 libdispatch.dylib              0x3a30 _dispatch_client_callout + 20
22 libdispatch.dylib              0x6eec _dispatch_continuation_pop + 500
23 libdispatch.dylib              0x6558 _dispatch_async_redirect_invoke + 584
24 libdispatch.dylib              0x15164 _dispatch_root_queue_drain + 396
25 libdispatch.dylib              0x1596c _dispatch_worker_thread2 + 164
26 libsystem_pthread.dylib        0x1080 _pthread_wqthread + 228
27 libsystem_pthread.dylib        0xe5c start_wqthread + 8
```

btw. shout-out to @ernesto-elsaesser for creating this library. It's a nice little util and it has served me well in the last years 🙏

**Comment by ernesto-elsaesser** [comment]
Thanks for the shoutout! :blush:

It looks like this happens deep inside the libxml2 buffer allocation code. Most likely the new OS generation broke some compatibility with libxml2, which is something Apple themselves had to fix. It might also be an issue with the way [Fuzi](https://github.com/cezheng/Fuzi) calls libxml2, but that package hasn't been touched in the last 4 years and simply calls the libxml2 top-level API, so that seems unlikely.

However, looking at the code again, I think the HTML reference extraction should also be doable with some regex magic, just like for CSS. I am currently setting up Xcode 16 to do some testing, but I might just throw out the XML parsing all together, as waiting for Apple to straighten out their release is no fun :)

Could you quickly check if the HTML documents of the URLs you are trying to archive are especially large? Do you have a specific URL that tends to crash?

**Comment by ernesto-elsaesser** [comment]
I just released version 1.1.0 using `XMLParser` from Foundation instead of Fuzi / libxml2. Please let me know if that fixes the issue. I could not reproduce the Xcode 16 crash myself.

**Comment by TheMen4ce** [comment]
Hey thanks for looking into the issue 😃

✅ I ran the app with release 1.1.0 and there are no crashes. When switching back to 1.0.1 the crashes re-appear.

❌ Unfortunately, archiving images no longer works. I've looked at your changes but I couldn't identify what could have gone wrong. 

Even with 1.0.1 the WebArchiver had "issues" when working with images that contain a `srcset`. It just archived the image that was in the `src` attribute but if the browser then would request another one, the image wasn't found (obviously). I worked around this by ensuring that the image linked in the `src`-tag matches the one the browser would request on a phone 😅 I guess fixing this would require the archiver to check first which image would be chosen by Safari (given Safari version and viewport size) and then archive this one.

Example:
```
<picture>
    <source srcset="/content/images/size/w600/format/webp/2024/09/asdf.jpeg 600w,
                /content/images/size/w1400/format/webp/2024/09/asdf.jpeg 1400w" 
sizes="(max-width: 600px) 100vw, (max-width: 992px) 85vw, (max-width: 1200px) 70vw, (max-width: 1828px) 47vw, 870px" type="image/webp">
    
<img srcset="/content/images/size/w600/2024/09/asdf.jpeg 600w,
                /content/images/size/w1400/2024/09/asdf.jpeg 1400w" sizes="(max-width: 600px) 100vw, (max-width: 992px) 85vw, (max-width: 1200px) 70vw, (max-width: 1828px) 47vw, 870px" 
src="/content/images/size/w1400/format/webp/2024/09/asdf.jpeg" >
</picture>
```

**Comment by TheMen4ce** [comment]
@ernesto-elsaesser I've updated on which platforms the error occurs:

ERROR: With Xcode 16, release build config, using an iPhone X with iOS 16.7.10.

NO ERROR with either: 
- release building with Xcode 15.4
- OR using the debug build config
- OR when using an iPhone 16 with iOS 18.0.0

**Comment by ernesto-elsaesser** [comment]
OK, so it was the combination of a new Xcode and an older phone / OS version. Something that Apple might not even care about, so dropping libxml2 was probably the right choice.

Regarding `srcset` attributes, that already goes into the direction of dynamic webpages. As I tried to explain in the README, attempting to trace all the possible resources a dynamic / responsive page might load just through static analysis of the sources is really not feasible, and this package certainly doesn't even try, but limits itself to the three very basics:

- `<link rel="stylesheet" href="...">`
- `<img src="...">`
- `<script src="...">`

See https://github.com/ernesto-elsaesser/WebArchiver/blob/master/Sources/WebArchiver/ReferenceCollector.swift#L20.

However, it should be quite easy to extend the new `ReferenceCollector` to also parse `srcset` attributes, providing additional references of type image for each listed size. If you don't mind having a "thick" webarchive with resources for different clients. But notice that this is a first step towards replicating the backend 😃 So I think your workaround of setting the `src` to the right-sized version for the requesting user agent is a sensible solution here, serving a more static page for archiving.

If you really choose to extend the code, of course feel free to create a PR! I just can't support specific requirements like that from my side, sorry.

**Comment by TheMen4ce** [comment]
No need to support `srcset` :) I just mentioned it since we were speaking about images not being loaded.

I'm not sure if we've understood each other [here](https://github.com/ernesto-elsaesser/WebArchiver/issues/3#issuecomment-2381437041):

WebArchiver 1.1.0:
✅ No crash on older iPhones/iOS when built with iOS 16 release built
❌ Doesn't archive images from `src` tag on all platforms

WebArchiver 1.0.1:
❌ Crashes on older iPhones/iOS when built with Xcode 16 release built
✅ Archives images from `src` tag on all platforms
✅ Works with Xcode 15.4 release build on all platforms
✅ Works with Xcode 16 release build on device with iOS 18

Older iPhones/iOS:
I got this error in production with 5765 crash events affecting 1393 users on devices up to iPhone 14. The devices were using iOS version up to 17 (most of them were on iOS 17.6.1). So it's not such an edge case 😅 The only reason we didn't get more crash events is cause I've published a fix release right after the build went live. 
I just use an old iPhone X to reproduce the issue 🥲

**Comment by cifilter** [comment]
Just chiming in to say we're experiencing similar issues with iOS 18 builds on _device only_ using `libxml2`. We don't crash, but we frequently get empty documents (but not 100% of the time!):
```
CCHARS COUNT: 642514
BUFFER COUNT: 642514
:1: HTML parser error : Document is empty

^
```

This thread seems to confirm our suspicion that something fundamental changed somewhere inside Apple's lib(s), and it `libxml2` can't read the memory layout of the HTML string (converted to a C string) anymore. 🤷‍♂️

**Comment by cifilter** [comment]
Found out that calling `Fuzi.XMLDocument(string:encoding:)` (same with `HTMLDocument`) results in parser failure (or maybe even crashes, as indicated in this crash report). _However_, calling `Fuzi.XMLDocument(data:)` instead resulted in success every time for us.

I notice this in the stack trace posted above:
`XMLDocument.__allocating_init(string:encoding:) + 130 (Document.swift:130)`

Might be worth switching over your call sites to just something like `XMLDocument(data: Data(xmlString.utf8))` to see what happens.

**Comment by ernesto-elsaesser** [comment]
Thank you both for the additional details and sorry for taking so long to get back to you.

@TheMen4ce I checked the issues with images you described and found that `XMLParser` aborts parsing as soon as it encounters any invalid XML ... which is basically all HTML. That might have been the reason why I used `Fuzi` in the first place ... Anyways, with the great finding of @cifilter I just published version [1.2.0](https://github.com/ernesto-elsaesser/WebArchiver/releases/tag/1.2.0), which again uses `Fuzi` to parse HTML, but using the safe `HTMLDocument(data:)` constructor. I hope this version now works in every scenario. Sorry for breaking your apps.
